internal/util/types.go

```diff
@@ -16,10 +16,49 @@ limitations under the License.
 
 package util
 
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/manager/scheme"
+)
+
 // Endpoint describes a kubernetes endpoint, same as a target in Kong.
 type Endpoint struct {
 	// Address IP address of the endpoint
 	Address string `json:"address"`
 	// Port number of the TCP port
 	Port string `json:"port"`
 }
+
+// TypeMeta is stripped after unmarshaling into Go struct due to the issue described in
+// https://github.com/kubernetes/kubernetes/issues/3030, but we need it for various purposes.
+
+// This function is adopted from https://github.com/kubernetes/cli-runtime/blob/v0.28.2/pkg/printers/typesetter.go#L39-L72
+// It has been modified to remove the io.Writer output and to use the scheme package directly instead of the Typer helper.
+
+// PopulateTypeMeta adds GVK information to a runtime.Object that may not have it available in the object TypeMeta.
+func PopulateTypeMeta(obj runtime.Object) error {
```

Comment:
Is it possible to add unit tests to `PopulateTypeMeta`?

Reply:
e5a4a42edb0df920a751d9af6d0b1c1eeb511e06